handlers.go

```diff
@@ -3153,6 +3155,66 @@ func (s *Server) handleAddContentsToCollection(c echo.Context, u *User) error {
 	return c.JSON(200, map[string]string{})
 }
 
+// handleCommitCollection godoc
+// @Summary      Produce a CID of the collection contents
+// @Description  This endpoint is used to save the contents in a collection, producing a top-level CID that references all the current CIDs in the collection.
+// @Param        coluuid     path     string  true     "coluuid"
+// @Tags         collections
+// @Produce      json
+// @Success      200  {object}  string
+// @Router       /collections/commit/{collection-id} [post]
+func (s *Server) handleCommitCollection(c echo.Context, u *User) error {
+	colid := c.Param("coluuid")
+
+	var col Collection
+	if err := s.DB.First(&col, "uuid = ? and user_id = ?", colid, u.ID).Error; err != nil {
+		return err
+	}
+
+	contents := []Content{}
+	if err := s.DB.Debug().
+		Model(CollectionRef{}).
+		Where("collection = ?", col.ID).
+		Joins("left join contents on contents.id = collection_refs.content").
+		Select("contents.*").
+		Scan(&contents).Error; err != nil {
+		return err
+	}
+
+	bserv := blockservice.New(s.Node.Blockstore, offline.Exchange(s.Node.Blockstore))
+	dserv := merkledag.NewDAGService(bserv)
+
+	collectionNode := merkledag.ProtoNode{}
+	for _, content := range contents {
+		curCID := content.Cid.CID
+		nd, err := dserv.Get(context.Background(), curCID)
+		if err != nil {
+			return err
+		}
+
+		lnk, err := format.MakeLink(nd)
+		if err != nil {
+			return err
+		}
+		collectionNode.AddRawLink("a link", lnk)
+	}
+
+	// update DB with new collection CID
+	col.CID = collectionNode.Cid().String()
+	if err := s.DB.Save(&col).Error; err != nil {
+		return err
+	}
+
+	ctx := c.Request().Context()
+	dserv.Add(ctx, &collectionNode)
+	pinstatus, err := s.CM.pinContent(ctx, u.ID, collectionNode.Cid(), collectionNode.Cid().String(), []*CollectionRef{}, []peer.AddrInfo{}, 0, nil)
```

Comment:
Regarding pinning, a side effect of this is that a `content` will be created for the collection node, and a deal will be attempted - is this what we want? that is, the contents in the collection will have their own deal, and the collection node (which will now be treated as a content) will have its own too

Reply:
Also, you need to get the host `AddrInfo` in case the pin request is sent to a remote shuttle, it will need the address to find the blocks with bitswap. In the case where the pin is done locally, I am not sure what happens when the host dial itself

Reply:
>In the case where the pin is done locally, I am not sure what happens when the host dial itself

I just checked the libp2p codebase, if we try to connect to self, it will fail with a `ErrDialToSelf` error. We need to do a check at https://github.com/application-research/estuary/blob/master/pinning.go#L102 to only connect if 

```
s.Node.Host.ID != pi.ID
```

the same check should be done https://github.com/application-research/estuary/blob/master/cmd/estuary-shuttle/main.go#L1373. We should have this checks in place even without this PR

cc @jimmylee @whyrusleeping

Reply:
We definitely don't want the nodes to make deals out of these. I'm unsure if `pinContent` actually fires up something that makes the deals. Can you point me to the code that actually does this? If that's really the case we can always manually pin the CID without adding it to the `contents` table

Reply:
It first creates a content here https://github.com/application-research/estuary/blob/master/pinning.go#L198 and then, for local content, it is added to the pinning queue here https://github.com/application-research/estuary/blob/master/pinning.go#L248 and the doPining operation queues it for a deal here https://github.com/application-research/estuary/blob/master/pinning.go#L116